### ISSUE TITLE
About "Native Implemention"

### ISSUE THREAD

**Issue by JUGGHM** [NVlabs/pacnet#22]
Hi Suhang! Thanks for your sharing the code! I just wonder why there is a native implemention to control the computational flow. Is there any theoretical or performance differences between the native implemantion and non-native implemention?

**Comment by suhangpro** [comment]
The native implementation is using only standard pytorch layers/operations, while the other one uses the Function interface at places. The latter can have some memory advantage for larger input sizes.

**Comment by Jerrypiglet** [comment]
Hey Hang, thanks for the comment! Additionally I am wondering if you have any statistics on how big the performance gap is between native and functional implementations, especially in case of large input sizes, or any differences in backprop speed, because I did some basic benchmarking and did not find significant difference there. Thanks!

**Comment by suhangpro** [comment]
There shouldn't be any noticeable differences in speed. The non-native implementation has advantages in terms of _peak memory usage_, but I have no statistics regarding the gap. The native version was there mostly for debugging purposes, and the non-native version should be the preferred option now.

**Comment by Jerrypiglet** [comment]
Ah i see. Thanks! The native version would also be useful for prototyping because it is easier to build upon.

BTW one more question though, in https://github.com/NVlabs/pacnet/blob/master/pac.py#L322, should it be:
```
ctx.save_for_backward(input if ctx.needs_input_grad[0] else None,
                              kernel if ctx.needs_input_grad[1] else None)
```
instead of 
```
ctx.save_for_backward(input if ctx.needs_input_grad[1] else None,
                              kernel if ctx.needs_input_grad[0] else None)
```
, given L339 and L351 for inference.

**Comment by suhangpro** [comment]
You are right. It's fixed now. Thanks!

**Comment by Jerrypiglet** [comment]
Hey Hang, after a second thought, the original version is actually correct to me. `kernel` will be used when you compute `grad_input`, and `input` will be used when you compute `grad_kernel`. See the highlights below.
![image](https://user-images.githubusercontent.com/4903961/101193372-7b4b7080-362a-11eb-9154-c5d40c605aec.png)

In your updated version I suppose you will see errors in https://github.com/NVlabs/pacnet/blob/master/pac.py#L352 that `input` is None when `ctx.needs_input_grad[0] == False` and `ctx.needs_input_grad[0] == True`. Am I right?

**Comment by suhangpro** [comment]
@Jerrypiglet That's right. I wasn't careful enough on this. The change is now reverted.
